src/modem.rs

```diff
@@ -0,0 +1,335 @@
+use at_commands::{builder::CommandBuilder, parser::CommandParser};
```

Comment:
![image](https://github.com/user-attachments/assets/665a4216-a3f5-4b1a-bd84-0b315f647573)

This is how I understand the `SimModem` will work. I want to be able to keep track of which mode the device is in using command mode and data mode.

I also want to provide the user code access to the `Read` and `Write` interface so that they can hook up the ppp `rx` and `tx` through `SimModem`

I think the `SimModem` will also be generic over type `T`, where `T: DCE`. DCE will be a new introduced trait, with a single impl for `SimComA7600`, which takes care of getting the device into data mode. I also think we should allow for `send_cmd` to be public so that you can still do things like:
get the iccid,
get the imei
get the signal strength

Reply:
> ![image](https://private-user-images.githubusercontent.com/16350310/359004889-665a4216-a3f5-4b1a-bd84-0b315f647573.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQwNzIxNjYsIm5iZiI6MTcyNDA3MTg2NiwicGF0aCI6Ii8xNjM1MDMxMC8zNTkwMDQ4ODktNjY1YTQyMTYtYTNmNS00YjFhLWJkODQtMGIzMTVmNjQ3NTczLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODE5VDEyNTEwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk2MmFiNWRiYmViYzZlM2MyZjNmNDVhNDY1NWFkNjEyNjBiY2RlY2I5NTBjOGFlMjI1MzBlZmJjNjE3MWEzOTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.2sNgvczothLr7Ul8QqERqYfUKXmviYzjvC0ghO3_l1s)

Not really sure what is "DTE" on that picture? For your modem, the "terminal" is just the pair of `embedded_io::Read + embedded_io::Write`. Why do you need anything else here?

> I also want to provide the user code access to the `Read` and `Write` interface so that they can hook up the ppp `rx` and `tx` through `SimModem`

I do not understand this. It is the **user code** that will give **your** sim modem the `Read` + `Write` traits, so the user code **apriori** has access to those. You only need to do something special if/when you plan to support MUX mode, where somehow on top of this user-provided Read+Write you'll implement two "virtual" tunnels:
- One, for commands, but you keep this inside the modem
- Another, for PPP (data) -> this you need to expose to the users

But maybe we need to think about this once you have support for MUX? Because at the moment, in the context of "either-commands-or-data" you don't need to provide to the user anything...

> I think the `SimModem` will also be generic over type `T`, where `T: DCE`. DCE will be a new introduced trait, with a single impl for `SimComA7600`, which takes care of getting the device into data mode.

Isn't this the same as my `PppRunner` thing?
By the way, I see you are copying the terminology from the C ESP modem. No offense, but this terminology is highly confusing... DTE, DCE... yeah.

> I also think we should allow for `send_cmd` to be public so that you can still do things like: get the iccid, get the imei get the signal strength

Sure no prob.

Reply:
> Isn't this the same as my PppRunner thing?
> By the way, I see you are copying the terminology from the C ESP modem. No offense, but this terminology is highly confusing... DTE, DCE... yeah.

No not quite. `PPPRunner` describes how to feed the `tx` and `rx`. `ModemDevice`(the C api calls this `DCE`) describes how to get a specific device into data mode. For this case `SIM7600` sends those specific commands, but not every modem will have the same formatting, delimiters, responses or newlines etc.

`DTE` is any "terminal", but essentially a device that implements `embedded_io::Read+embedded_io::Write`

Reply:
Ah, ok. Instead of having multiple `SimModem`s, like `struct Sim7600Modem`, `struct Br96Modem`, `struct Sim800Modem` where these all implement a common `trait SimModem` trait, you want to model it slightly differently.

OK.

Reply:
I opted to keep it simpler. `SimModem` is a trait with 
`negotiate()` and `get_mode()`. Can probably add a few more general  things such as signal strength, ICCID, IMEI.

`EspModem` stays as a kind of glue between `EspNetifDriver` and `UartDriver`